geneologijos/geneologijos.pde

```diff
@@ -283,12 +243,82 @@ void loadJSON(File selection){
         textboxes.add(new Textbox(new PVector(px,py)));
         String  txt = item.getString("text");
         textboxes.get(textboxes.size()-1).text = txt;
+        textboxes.get(textboxes.size()-1).urselected = false;
       }
     }catch (Exception e){
       println("LOADING FAILED");
     }
   } 
 }
 void saveJSON(File file){
-  level.save_level(file.getAbsolutePath());
+  level.save_level(file.getAbsolutePath()+".geneo");
+}
+void menu_phases(){
+  if (menuPhase==1) {
+    textSize(60);
+    fill(0);
+    textAlign(LEFT, BOTTOM);
+    text("GYVAS", (selected_tile.x+1)*tileSize+offx, selected_tile.y*tileSize+offy);
+    textAlign(LEFT, TOP);
+    text("NEGYVAS", (selected_tile.x+1)*tileSize+offx, selected_tile.y*tileSize+offy);
+    if (click) {
+      click=false;
+      if (mouseX>(selected_tile.x+1)*tileSize+offx&&mouseX<(selected_tile.x+1)*tileSize+offx+textWidth("GYVAS")) {
+        if (mouseY>selected_tile.y*tileSize+offy-60&&mouseY<selected_tile.y*tileSize+offy) {
+          level.tiles[(int)selected_tile.x][(int)selected_tile.y].dead=false;
+          menuPhase++;
+        } else if (mouseY>selected_tile.y*tileSize+offy&&mouseY<selected_tile.y*tileSize+offy+60) {
+          level.tiles[(int)selected_tile.x][(int)selected_tile.y].dead=true;
+          menuPhase++;
+        } else {
+          menuPhase=0;
+        }
+      } else {
+        menuPhase=0;
+      }
+    }
+  }
+  if (menuPhase==2) {

```

Comment:
This is just an illustration of the point because this happens all over the codebase really.

The code is difficult to read because you don't use any spacing, neither whitespace or new lines. 
1. You can use whitespace to separate variables from operators to make it easier to read. E.g. instead of `if (hovered_tile.x>selected_tile.x&&hovered_tile.x<selected_tile.x+img.length+1)` do `if (hovered_tile.x > selected_tile.x && hovered_tile.x < selected_tile.x + img.length + 1)`
2. More importantly using new lines to separate different "logical" parts of your code makes reading it _so_ much easier. Here I separate it into three parts: draw setup, drawing stuff, whatever the last bit is. 
```java
if (menuPhase==2) {
    fill(125);
    noStroke();
    
    rect((selected_tile.x+1)*tileSize+offx, selected_tile.y*tileSize+offy, img.length*tileSize, tileSize);
    for (int i = 0; i<img.length; i++) {
        image(img[i][0], (selected_tile.x+i+1)*tileSize+offx, (selected_tile.y)*tileSize+offy, tileSize, tileSize);
    }
    
    if (click&&mouseButton==LEFT) {
        click=false;
        if (hovered_tile.y==selected_tile.y) {
            if (hovered_tile.x>selected_tile.x&&hovered_tile.x<selected_tile.x+img.length+1) {
                menuPhase++;
                level.tiles[(int)selected_tile.x][(int)selected_tile.y].imgy=-1;
                level.tiles[(int)selected_tile.x][(int)selected_tile.y].imgx=int(hovered_tile.x-selected_tile.x-1);
            } else {
                menuPhase=0;
            }
        } else {
            menuPhase=0;
        }
    }
}
```

Combining these two points give us:
```java
if (menuPhase == 2) {
    fill(125);
    noStroke();
    
    rect((selected_tile.x + 1) * tileSize + offx, selected_tile.y * tileSize + offy, img.length * tileSize, tileSize);
    for (int i = 0; i < img.length; i++) {
        image(img[i][0], (selected_tile.x + i + 1) * tileSize + offx, selected_tile.y * tileSize + offy, tileSize, tileSize);
    }
    
    if (click && mouseButton == LEFT) {
        click = false;
        if (hovered_tile.y == selected_tile.y) {
            if (hovered_tile.x > selected_tile.x && hovered_tile.x < selected_tile.x + img.length + 1) {
                menuPhase++;
                level.tiles[(int)selected_tile.x][(int)selected_tile.y].imgy = -1;
                level.tiles[(int)selected_tile.x][(int)selected_tile.y].imgx = int(hovered_tile.x - selected_tile.x - 1);
            } else {
                menuPhase = 0;
            }
        } else {
            menuPhase = 0;
        }
    }
}
```

And then finally if we do just a touch of refactoring we achieve the following:
```java
if (menuPhase == 2) {
    fill(125);
    noStroke();
    
    final borderX = (selected_tile.x + 1) * tileSize + offx;
    final borderY = selected_tile.y * tileSize + offy;
    rect(borderX, borderY, img.length * tileSize, tileSize);
    for (int i = 0; i < img.length; i++) {
        image(img[i][0], borderX + i * tileSize, borderY, tileSize, tileSize);
    }
    
    if (click && mouseButton == LEFT) {
        click = false;

        final hoveredSelectedY = hovered_tile.y == selected_tile.y;
        final hoveredWithinSelectedX = hovered_tile.x > selected_tile.x && hovered_tile.x < selected_tile.x + img.length + 1;
        if (hoveredSelectedY && hoveredWithinSelectedX) {
            menuPhase++;
            level.tiles[(int)selected_tile.x][(int)selected_tile.y].imgy = -1;
            level.tiles[(int)selected_tile.x][(int)selected_tile.y].imgx = int(hovered_tile.x - selected_tile.x - 1);
        } else {
            menuPhase = 0;
        }
    }
}
```